### ISSUE TITLE
[MNG-7106] VersionRange.toString() produces a string that cannot be parsed with VersionRange.createFromVersionSpec() for same lower and upper bounds

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#8342]
**[Akshay Shankara](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=akshay95)** opened **[MNG-7106](https://issues.apache.org/jira/browse/MNG-7106?redirect=false)** and commented

When a hard version requirement (Ex - [1.0]) is passed to [VersionRange#createFromVersionSpec(String)](https://github.com/apache/maven/blob/maven-3.6.3/maven-artifact/src/main/java/org/apache/maven/artifact/versioning/VersionRange.java#L106), it is parsed to [1.0, 1.0]. But, this version range is invalid [according to this exception](https://github.com/apache/maven/blob/maven-3.6.3/maven-artifact/src/main/java/org/apache/maven/artifact/versioning/VersionRange.java#L214-L217). If the parsed version range ([1.0, 1.0]) is converted to a String, it cannot be parsed once again by VersionRange#createFromVersionSpec(String).

Steps to reproduce -

```java
VersionRange versionRange = VersionRange.createFromVersionSpec("[1.0]"); // [1.0, 1.0]
String stringVersionRange = VersionRange.toString(versionRange); // "[1.0, 1.0]"
VersionRange exceptionVersionRange = VersionRange.createFromVersionSpec(stringVersionRange); // <- InvalidVersionSpecificationException( "Range cannot have identical boundaries: [1.0, 1.0]" )
```

---

**Affects:** 3.6.3

**Issue Links:**
- [MNG-7561](https://issues.apache.org/jira/browse/MNG-7561) DefaultVersionRangeResolver should not try to resolve versions with same upper and lower bound

**Remote Links:**
- [GitHub Pull Request #825
](https://github.com/apache/maven/pull/825)
- [GitHub Pull Request #826
](https://github.com/apache/maven/pull/826)
- [GitHub Pull Request #922
](https://github.com/apache/maven/pull/922)

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1), [3.9.0](https://github.com/apache/maven/milestone/2?closed=1), [3.8.7](https://github.com/apache/maven/milestone/108?closed=1)

1 votes, 7 watchers

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

This sounds valid and inconsistent.

**Comment by jira-importer** [comment]
**[Brandon Heck](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brandon.heck)** commented

Would the preference be to fix this in parseRestriction by skipping the creation of a Restriction for a hard version requirement? Would it be preferable to fix this in the `toString` by checking the bounds for equality when constructing the string? Is there another way this might be fixed that would be better than either of those suggestions? I'm assuming that the first suggestion may have unintended consequences when building the effective model.

**Comment by jira-importer** [comment]
**[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)** commented

This works fine already in Maven Resolver (verified in https://github.com/apache/maven-resolver/pull/202/files) so I think just dropping the exception when parsing string `"[1.0,1.0]"` is the correct solution here.

**Comment by jira-importer** [comment]
**[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)** commented

Fixed for 4.0.0 in https://github.com/apache/maven/commit/78edd431226204316e2d472d152b869a4a17ce61.
Fixed for 3.9.0 in https://github.com/apache/maven/commit/75e19406952a4e1a58b382ec5c68dd16c101670b

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin), is this bugfix worth back porting to 3.8.7?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [13d82830eed4d8a7ff1215fe559b041eb265bd7b](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=13d82830eed4d8a7ff1215fe559b041eb265bd7b) for `maven-3.8.x` branch.
